x-pack/metricbeat/module/sql/query/query.go

```diff
@@ -81,167 +79,81 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
-	}
-	return m.db, nil
-}
-
-// fetchTableMode scan the rows and publishes the event for querys that return the response in a table format.
-func (m *MetricSet) fetchTableMode(rows *sqlx.Rows, report mb.ReporterV2) error {
-
-	// Extracted from
-	// https://stackoverflow.com/questions/23507531/is-golangs-sql-package-incapable-of-ad-hoc-exploratory-queries/23507765#23507765
-	cols, err := rows.Columns()
-	if err != nil {
-		return errors.Wrap(err, "error getting columns")
-	}
-
-	for k, v := range cols {
-		cols[k] = strings.ToLower(v)
-	}
-
-	vals := make([]interface{}, len(cols))
-	for i := 0; i < len(cols); i++ {
-		vals[i] = new(interface{})
-	}
-
-	for rows.Next() {
-		err = rows.Scan(vals...)
-		if err != nil {
-			m.Logger().Debug(errors.Wrap(err, "error trying to scan rows"))
-			continue
-		}
-
-		numericMetrics := common.MapStr{}
-		stringMetrics := common.MapStr{}
-
-		for i := 0; i < len(vals); i++ {
-			value := getValue(vals[i].(*interface{}))
-			num, err := strconv.ParseFloat(value, 64)
-			if err == nil {
-				numericMetrics[cols[i]] = num
-			} else {
-				stringMetrics[cols[i]] = value
-			}
-
+		for _, ms := range mss {
+			report.Event(m.getEvent(ms))
 		}
 
-		report.Event(mb.Event{
-			RootFields: common.MapStr{
-				"sql": common.MapStr{
-					"driver": m.Driver,
-					"query":  m.Query,
-					"metrics": common.MapStr{
-						"numeric": numericMetrics,
-						"string":  stringMetrics,
-					},
-				},
-			},
-		})
+		return nil
 	}
 
-	if err = rows.Err(); err != nil {
-		m.Logger().Debug(errors.Wrap(err, "error trying to read rows"))
+	ms, err := db.FetchVariableMode(ctx, m.Query)
+	if err != nil {
+		return err
 	}
+	report.Event(m.getEvent(ms))
 
 	return nil
 }
 
-// fetchVariableMode scan the rows and publishes the event for querys that return the response in a key/value format.
-func (m *MetricSet) fetchVariableMode(rows *sqlx.Rows, report mb.ReporterV2) error {
-	data := common.MapStr{}
-	for rows.Next() {
-		var key string
-		var val interface{}
-		err := rows.Scan(&key, &val)
-		if err != nil {
-			m.Logger().Debug(errors.Wrap(err, "error trying to scan rows"))
-			continue
-		}
-
-		key = strings.ToLower(key)
-		data[key] = val
+func (m *MetricSet) getEvent(ms common.MapStr) mb.Event {
+	return mb.Event{
+		RootFields: common.MapStr{
+			"sql": common.MapStr{
+				"driver":  m.Driver,
+				"query":   m.Query,
+				"metrics": getMetrics(ms),
+			},
+		},
 	}
+}
 
-	if err := rows.Err(); err != nil {
-		m.Logger().Debug(errors.Wrap(err, "error trying to read rows"))
-	}
+func getMetrics(ms common.MapStr) (ret common.MapStr) {
+	ret = common.MapStr{}
 
 	numericMetrics := common.MapStr{}
 	stringMetrics := common.MapStr{}
-
-	for key, value := range data {
-		value := getValue(&value)
-		num, err := strconv.ParseFloat(value, 64)
-		if err == nil {
-			numericMetrics[key] = num
-		} else {
-			stringMetrics[key] = value
+	boolMetrics := common.MapStr{}
+
+	for k, v := range ms {
+		switch v.(type) {
+		case float64:
```

Comment:
Are all numeric values typed as `float64`?

Reply:
I forgot to answer this. I spent quite a lot of time with this: The issue is that sql drivers handle numeric type inconsistently between them and the generic result is `float64`. I mean that while driver A handles `INTEGER` sql type properly and returns `int32` go types, driver B handles all `INTEGER` sql type as `float64`. Same problem with `NUMERIC`.

Sad story here is that Go's SQL driver has the tools to deal with this properly https://golang.org/src/database/sql/driver/driver.go?s=15846:15947#L413 but not all drivers implement it

Reply:
So is `float64` enough here or we should add some other numeric type to be covered?

Reply:
Support for specific numeric types is out of the scope of this PR and it's not supported by the original module either. We can open an issue later to discuss about complexity and benefits first.